Scrub credentials in LDAP bind requests.

Ayanda-D opened the pull request:
Fixes #12.

michaelklishin commented on src/rabbit_auth_backend_ldap.erl:
```diff
@@ -30,6 +30,7 @@
 -define(L(F, A),  log("LDAP "         ++ F, A)).
 -define(L1(F, A), log("    LDAP "     ++ F, A)).
 -define(L2(F, A), log("        LDAP " ++ F, A)).
+-define(HIDDEN_CRED,  "xxxx").
```
"Hidden" makes me think that we are doing something shady. Can this be renamed to "SCRUBBED"? And lets expand the "cred" while we are at it.

michaelklishin commented on src/rabbit_auth_backend_ldap.erl:
```diff
@@ -504,6 +507,56 @@ creds(#auth_user{impl = #impl{user_dn = UserDN, password = PW}}, as_user) ->
     {ok, Creds}.
 
+%% Maybe hide credentials
+maybe_hide_creds([], Acc)      -> lists:reverse(Acc);
```
Same as above, I'd use "scrub" instead of "hide."

Ayanda-D commented:
Done with some minor suggested updates.

hairyhum commented on src/rabbit_auth_backend_ldap.erl:
```diff
@@ -504,6 +507,56 @@ creds(#auth_user{impl = #impl{user_dn = UserDN, password = PW}}, as_user) ->
+maybe_scrub_creds([], Acc)      -> lists:reverse(Acc);
+maybe_scrub_creds([H|Rem], Acc) -> maybe_scrub_creds(Rem, [maybe_scrub_cred(H)|Acc]).
+
+maybe_scrub_cred({'BindRequest', N, DN, {simple, PWD}}) ->
```
If it scrubs only `BindRequest`, should setting be called `scrub_bind_request_credentials`?
Because it doesn't seem like not logging bind credentials.
`LDAP bind succeeded: cn=Alice,ou=People,dc=example,dc=com` can look like logging credentials for user.

Ayanda-D replied:
@hairyhum yes, you're right. i'm updating this. thanks.

Ayanda-D replied:
I'm just retaining the `maybe_` prefix, as scrubbing is dependant on config. 👍

hairyhum commented:
I think we should enable this setting by default and/or mention it in `log` setting documentation.
Right now it says that `network` setting is unsafe and can expose passwords.

Ayanda-D commented:
Okay cool. Yeah makes sense. Will enable & update notes 👍

Ayanda-D commented:
@hairyhum thanks, config now enabled by default. i've also just added a test for when disabled, and updated the [docs](https://github.com/rabbitmq/rabbitmq-website/commit/0913c9a26548dc117b5143e11a9068cd5400e8c5).

hairyhum commented:
This implementation make `scrub_bind_request_credentials` config option only applicable if `log` is set to `network`. It doesn't look clear for me and can confuse users.
@michaelklishin @Ayanda-D any thoughts?

michaelklishin commented:
Do we currently log credentials with other log levels? If we do then the option should cover them, too; otherwise it's OK with me.

Ayanda-D commented:
Depends on what we consider `credentials` to be. If `DN` on its own, without the password is sensitive info, and also needs to be scrubbed, then yes, adding this on both log levels would make sense. Otherwise we keep this under `network` logs, when credentials are inclusive of password.

hairyhum commented:
In case we want to scrub network only logs, there can be additional option for `log` config (e.g. `network_unsafe`), instead of separate configuration key.

michaelklishin commented:
That's an interesting idea. Scrubbing by default and logging everything (including credentials) in a separate log level sounds good to me.

Ayanda-D commented:
+1. yeah, that sounds much better. Okay, looking into this. I'll ping you guys once its ready. 👍

Ayanda-D committed:
Adds functionality to scrub credentials in LDAP network
traffic logs, for bind requests.

Ayanda-D committed:
Introduces logging tests, and tests for scrubbing
bind request credentials in network traffic logs.

Ayanda-D committed:
Introduces and adds handling of 'network_unsafe' log type
option.

Ayanda-D committed:
Adds logging test for 'network_unsafe' log option.

Ayanda-D commented:
@michaelklishin @hairyhum this now scrubs credentials using the `network` log config, (even on failed bind requests), and `network_unsafe` option will keep them displayed. Please have a look/review. & the [vocab](https://github.com/rabbitmq/rabbitmq-website/compare/rabbitmq-auth-backend-ldap-12) I've used in the docs. Thanks for your input!

hairyhum commented on src/rabbit_auth_backend_ldap.erl:
```diff
@@ -314,14 +323,16 @@ with_login(Creds, Servers, Opts, Fun) ->
                     case eldap:simple_bind(LDAP, UserDN, Password) of
                         ok ->
-                            ?L1("bind succeeded: ~s", [UserDN]),
+                            ?L1("bind succeeded: ~s",
```
This log entry will be scubbed with `network`, but not with `true` log verbosity. This doesn't make mush sense to me.

Ayanda-D replied:
@hairyhum Okay cool. I'm extending this to also be scrubbed under `true` log verbosity. So ONLY under `network_unsafe` would credentials be exposed, and in the minimal LDAP decision logs, we'd still have credentials/DN in that bind outcome log entry, scrubbed by default. Makes sense. Thanks! 👍

Ayanda-D committed:
Extends scrubbing of bind request credentials to be
carried out on all log levels, except 'network_unsafe'.

Ayanda-D commented:
@hairyhum thanks, i've done the update. please have a look.

Ayanda-D committed:
Ignore scrubbing operations for 'false' log level.
